Document symex instructions

NlightNFotis opened the pull request:
Add documentation for the `goto-symex` and `goto-program-instructions`.

NlightNFotis committed:
Add some high-level documentation on `symex` and GOTO-program instructions.

martin-cs commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,91 @@
+The `goto_modelt` is the main data structure that CBMC (and the other tools) use to
+represent GOTO-IR (the `GOTO` Intermediate Representation).
+
+A `goto_modelt` is effectively a product type, consisting of:
```
Same comment as on the previous draft; I would suggest using "pair" or "tuple" as a more C++-ish way of saying this.

martin-cs commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,91 @@
+
+A `goto_modelt` is effectively a product type, consisting of:
+
+* A list of GOTO functions (pseudocode: `type goto_functionst = list<goto_functiont>`)
```
It's not really a list...

martin-cs commented on doc/architectural/central-data-structures.md:
````diff
@@ -0,0 +1,91 @@
+A `goto_functiont` is also a product type. It's designed to represent a function
+at the IR level, and effectively it's the following ADT (in pseudocode):
+
+```js
````
Could you use C++-style pseudocode?

martin-cs replied:
And / or a reference to the code.

martin-cs commented on doc/architectural/central-data-structures.md:
```diff
@@ -0,0 +1,91 @@
+Another important data structure that needs to be discussed at this point is
+`source_locationt`.
+
+This is an `irept`. `irep`s are the central data structure that model most entities inside
```
`irept` is pretty critical and deserves space of its own, not just under `source_locationt`

jimgrundy replied:
+1

martin-cs commented on doc/architectural/symex-instructions.md:
```diff
@@ -0,0 +1,127 @@
+
+## A (very) short introduction to Symex
+
+Symex is, at its core, a GOTO-program interpreter. While Symex is interpreting
```
'an interpreter BUT using symbolic values, not literal ones. For example...  Note that this produces a formula which describes all possible outputs rather than a single output value.'

esteffin replied:
Agreed. Changed

martin-cs commented on doc/architectural/symex-instructions.md:
````diff
@@ -0,0 +1,127 @@
+}
+```
+
+If we ask CBMC to print the GOTO-program instructions with `--show-goto-functions`, then part
````
Should be in the `goto-functions` documentation rather than just in symex.

jimgrundy replied:
I presume (hope) it is. Maybe here we should have a cross reference to that part of the goto-functions documentation.

martin-cs commented on regression/cbmc/field-sensitivity15/test.desc:
```diff
@@ -0,0 +1,9 @@
+^SIGNAL=0$
+--
+--
+The test is a minimized version of code generated from s2n-quic using Kani.
```
More tests are always good but why is this in a documentation PR?  I guess I have missed something...

martin-cs commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -215,14 +215,6 @@ void symex_assignt::assign_non_struct_symbol(
       assignment.rhs,
       target,
       symex_config.allow_pointer_unsoundness);
-    // Erase the composite symbol from our working state. Note that we need to
```
There could well be a good justification for this removal of code but please could they not be in a documentation PR?

NlightNFotis commented:
Hi @martin-cs , this was branched of at an earlier point from the same branch as the other documentation PR (#7470), and it hasn't tracked the changes that addressed the comments on the other PR, so it appears like it has the same problems, but it's just being stale.

This means that the earlier commits here are just adding noise, so I will focus on addressing everything on the other PR, get it merged, and then rebase this one on top of `develop`, for a cleaner, less stale history. I will then address the comments that remain relevant after the rebase and ask you for a re-review.

jimgrundy commented on doc/architectural/cbmc-architecture.md:
```diff
@@ -64,9 +71,10 @@ To be documented.
 To be documented.
 
-## Goto functions &rarr; BMC &rarr; Counterexample (trace) ##
+## Goto functions &rarr; BMC &rarr; Counterexample (trace)
```
Why this change? Doesn't seem to match the other lines.

NlightNFotis replied:
I did this as a drive-by clean up. As far as I know headings in markdown are only prepended by a `##`.

NlightNFotis replied:
On a second visit, it appears that [doxygen does support any number of `#`](https://doxygen.nl/manual/markdown.html) after the title, but that's invalid markdown.

If this change offends in any capacity however, it's easy to be reverted.

tautschnig replied:
It seems this file has that issue multiple times. Can you please clean up all occurrences? (src/goto-instrument/README.md also suffers from this problem. Maybe do one commit that does the cleanup for all of these?)

jimgrundy replied:
I mean fair enough. I was thinking maybe there should be a separate PR that does this clean up. Or, at least, maybe consistently clean them all up in this file as there are many line this (as Michael as noticed). You want to leave the file in a consistent state so that newcomers will follow the example.

esteffin replied:
I will make trailing `#` consistent by removing them in the files touched by this PR.

jimgrundy commented on doc/architectural/symex-instructions.md:
````diff
@@ -0,0 +1,127 @@
+purposes:
+
+```c
+enum goto_program_instruction_typet
````
Maybe throw in some "..."s here to emphasize that this isn't the full set.

jimgrundy commented on doc/architectural/symex-instructions.md:
```diff
@@ -0,0 +1,127 @@
+enum goto_program_instruction_typet
+{
+  GOTO = 1,              // branch, possibly guarded
+  ASSUME = 2,            // non-failing guarded self loop
```
This is perhaps not the worlds most obvious description of an assumption.

esteffin replied:
The reason for this is that it's a copy from `goto_program_instruction_typet` where the comment of `ASSUME` states what was written.

I changed it to `ASSUME` to avoid confusion.

jimgrundy commented on doc/architectural/symex-instructions.md:
````diff
@@ -0,0 +1,127 @@
+```c
+int main(int argc, char **argv)
+{
+    int arry[] = {0, 1, 2, 3};
````
It this point you may as well just say "array", or go the whole way and call it "a". "arry" offers neither brevity nor clarity.

esteffin replied:
Changed to `a` and improved assertion comment

jimgrundy commented on doc/architectural/symex-instructions.md:
```diff
@@ -0,0 +1,127 @@
+dispatching functions are usually in their own file, designated by the instruction's
+name. As an example, you can find the code for the function goto_symext::symex_goto
+in [symex_goto.cpp](../../src/goto-symex/symex_goto.cpp)
+
```
A description of what symex produces, why, and what it means could use a longer explanation.

jimgrundy commented on src/goto-checker/multi_path_symex_only_checker.cpp:
```diff
@@ -74,8 +74,8 @@ void multi_path_symex_only_checkert::generate_equation()
 {
   const auto symex_start = std::chrono::steady_clock::now();
 
```
This PR seems like it should be a couple of PRs about different things. This doesn't feel like it belongs in a documentation PR.

tautschnig commented on doc/architectural/symex-instructions.md:
```diff
@@ -0,0 +1,127 @@
+
+## A (very) short introduction to Symex
+
+Symex is, at its core, a GOTO-program interpreter. While Symex is interpreting
+the program, it's also building a list of SSA steps that form part of the equation
+that is to be sent to the solver.
```
Would it be better to refer the reader to `src/goto-symex/README.md`?

esteffin replied:
Good idea, added

tautschnig commented on doc/architectural/symex-instructions.md:
````diff
@@ -0,0 +1,127 @@
+  // 3 file /tmp/example.c line 5 function main
+  DEAD main::1::arry
+  // 4 file /tmp/example.c line 5 function main
+  SET RETURN VALUE side_effect statement="nondet" is_nondet_nullable="1"
+  // 5 file /tmp/example.c line 5 function main
+  END_FUNCTION
+```
+
+In the above excerpt, the capitalised words at the beginning each instruction are the
+correspondent instruction types (`DECL`, `ASSIGN`, etc).
````
All of the above really belongs to goto programs, not so much to symex, does it?

esteffin replied:
Yes, but in this file we are describing the **Instruction type** that is exactly what we are showing.
In my opinion it is useful to show how the goto-program looks like.

Commit:
Removed trailing # from markdown

Commit:
Addressed some PR reviews comment

esteffin commented:
@martin-cs Can you please have a look and confirm your concerns have been addressed?

TGWDB commented:
@martin-cs Can you please act on this soon? It's been a week since the last request for review and more than that since various fixes were pushed. Note that we plan to force merge this in about 24 hours unless there is some new specific concern. We can make further changes in follow up tickets if necessary, but we'd like to merge this and progress ASAP.
